### PR TITLE
Movement fixes, refactor and unit test suite

### PR DESCRIPTION
**Pull request by PabloMansanet**
This started with the intention of adding unit tests to movement, but it fixes a handful of bad behaviours along the way, as well as refactor the module to be more functional.

To the best of my knowledge, after this change the word motions 100% reflect Kakoune, and edge cases are covered in the unit test suite.

### CODE REVIEW

File: helix-core/src/syntax.rs [hunk 1]
```diff
@@ -716,8 +716,11 @@ impl LanguageLayer {
 // For now cheat and just throw out non-root layers if they exist. This should still improve
 // parsing in majority of cases.
 
-use std::sync::atomic::{AtomicUsize, Ordering};
-use std::{iter, mem, ops, str, usize};
+use std::{
+    iter, mem, ops, str,
+    sync::atomic::{AtomicUsize, Ordering},
+    usize,
+};
```
Comment (pickfire) [inline helix-core/src/syntax.rs:723]: Why not separate them?

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -249,9 +228,9 @@ pub fn move_next_word_start(cx: &mut Context) {
     let (view, doc) = cx.current();
     let text = doc.text().slice(..);
 
-    let selection = doc.selection(view.id).transform(|range| {
-        movement::move_next_word_start(text, range.head, count).unwrap_or(range)
```
Comment (pickfire) [inline helix-term/src/commands.rs:253]: Why can we remove the `unwrap_or(range)`?

File: helix-term/src/ui/completion.rs [hunk 3]
```diff
@@ -11,8 +11,10 @@ use std::borrow::Cow;
 use helix_core::{Position, Transaction};
 use helix_view::Editor;
 
-use crate::commands;
-use crate::ui::{menu, Markdown, Menu, Popup, PromptEvent};
+use crate::{
+    commands,
+    ui::{menu, Markdown, Menu, Popup, PromptEvent},
+};
```
Comment (pickfire) [inline helix-term/src/ui/completion.rs:17]: Why not keep this?

File: helix-core/src/movement.rs [hunk 4]
```diff
@@ -1,184 +1,254 @@
...
-        }
-
-        skip_over_next(slice, &mut end, char::is_whitespace);
-    }
-
-    Some(Range::new(begin, end - 1))
+pub fn move_next_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Option<Range> {
+        let after_head = (range.head + 1).skip_newlines(slice);
+        (after_head + 1).inside(slice).then(|| {
```
Comment (pickfire) [inline helix-core/src/movement.rs:89]: I think `then` is a bit tool overkill, feel like we are doing code golf, it's harder to understand compared to `if cond { } else { }`. Also, what does the `?` at the end do?

File: helix-core/src/movement.rs [hunk 5]
```diff
@@ -1,184 +1,254 @@
...
+    fn outside(&self, slice: RopeSlice) -> bool {
+        !self.inside(slice)
+    }
 
-    for _ in 0..count {
-        if begin + 2 >= slice.len_chars() {
-            return None;
-        }
+    fn is_boundary(&self, slice: RopeSlice) -> bool {
+        (self + 1).inside(slice)
```
Comment (pickfire) [inline helix-core/src/movement.rs:165]: Why do we need this check when it will be called?

File: helix-core/src/movement.rs [hunk 6]
```diff
@@ -1,184 +1,254 @@
...
+    fn is_whitespace(&self, slice: RopeSlice) -> bool;
+    fn is_end_of_line(&self, slice: RopeSlice) -> bool;
+}
 
-        if is_word(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_word);
-        } else if is_punctuation(ch) {
-            with_end = skip_over_prev(slice, &mut end, is_punctuation);
-        }
+impl SliceIndexHelpers for usize {
```
Comment (pickfire) [inline helix-core/src/movement.rs:155]: I personally think this `impl` makes the code harder to understand since it's `usize` and we shouldn't `impl` it for `usize`, maybe a specific type yes.

File: helix-core/src/syntax.rs (hunk 1 above)
Comment (PabloMansanet) [reply helix-core/src/syntax.rs:723]: I briefly mentioned this in the channel: This is just the default `rustfmt` settings clashing with the ones that were used for the project thus far. I think it's a good idea if the formatting options for this project are commited in, so there's no noise in the diffs.

File: helix-term/src/ui/completion.rs (hunk 3 above)
Comment (PabloMansanet) [reply helix-term/src/ui/completion.rs:17]: Me too, again, it's the default behaviour `cargo fmt` overwriting what was there before; I don't know what options were used to autoformat the previous variant, but if someone has them I'm happy to re-run it.

File: helix-term/src/commands.rs (hunk 2 above)
Comment (PabloMansanet) [reply helix-term/src/commands.rs:253]: I had to refactor the API to ensure the `count` worked. Before, passing a high count so that one of the moves failed returned a `None`. Now, the `move` functions will return the latest achievable range after iterating until it's not possible to move anymore.

File: helix-core/src/movement.rs (hunk 4 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:89]: Personally, I find
```
if cond {
       Some(value)
   } else {
       None
   }
```

Wordier and less idiomatic than
```rust
condition.then(|| Some(value))
```
I don't feel there's any code golfing going on there, but I'm happy to subject it to consensus if other people weigh in :).

As for the `?`, it's needed to break out of the closure with a `None` when we've reached the end of the slice, and thus don't want to make any further anchor moves. This `?` is specifically what fixes the previous bug of moving the anchor when already having selected the last word.

File: helix-core/src/movement.rs (hunk 5 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:165]: In the next line, we're calling `categorize` on `self+1`, so we need to ensure it's within bounds.

File: helix-core/src/movement.rs (hunk 6 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:155]: Any particular reason why you wouldn't want to impl for `usize`? (Particularly for private traits like these).

File: helix-core/src/movement.rs (hunk 4 above)
Comment (kevinsjoberg) [reply helix-core/src/movement.rs:89]: I have fairly limited Rust knowledge at this point given I've only started learning a couple of days ago, but I do plenty of experience working with monads in other languages such as Haskell, Elm and the dry-monads in Ruby.

Personally, I agree with @PabloMansanet here. `.then` is common and I also tend to prefer it over `if/else`.

File: helix-term/src/ui/completion.rs (hunk 3 above)
Comment (archseer) [reply helix-term/src/ui/completion.rs:17]: We're just using out of the box cargo fmt settings. Are you sure you're not on nightly or have a global fmt config? There's an option to merge imports but it was too eager so it was disabled by default in fmt.

File: helix-core/src/movement.rs (hunk 4 above)
Comment (archseer) [reply helix-core/src/movement.rs:89]: I think `.then` is perfectly fine, it was only added recently to Rust so it's not as widely used yet.

File: helix-term/src/ui/completion.rs (hunk 3 above)
Comment (PabloMansanet) [reply helix-term/src/ui/completion.rs:17]: Ah, I think you caught me with nightly, that must be it! I'll run it from stable and retry

File: helix-core/src/movement.rs (hunk 6 above)
Comment (pickfire) [reply helix-core/src/movement.rs:155]: I find these parts made the upper parts harder to understand.

File: helix-core/src/movement.rs (hunk 5 above)
Comment (pickfire) [reply helix-core/src/movement.rs:165]: But it is called outside the function already so we probably don't need to check next line.

Comment (PabloMansanet) [reply helix-core/src/movement.rs:165]: These redundant checks are trivially optimized away by the compiler, so it's generally better to have the extra safety. If someone in the future reuses `is_boundary` for something else in this module, we are averting a panic or a silent error.

Comment (pickfire) [reply helix-core/src/movement.rs:165]: > These redundant checks are trivially optimized away by the compiler, so it's generally better to have the extra safety. If someone in the future reuses is_boundary for something else in this module, we are averting a panic or a silent error.

Are you sure about that?

File: helix-core/src/words.rs [hunk 7]
```diff
@@ -1,35 +1,19 @@
...
-    }
+pub fn nth_prev_word_boundary(slice: RopeSlice, index: usize, count: usize) -> usize {
+    (0..count).fold(index, |mut index, _| {
+        index = backwards_skip_while(slice, index, is_end_of_line).unwrap_or(index);
+        index = backwards_skip_while(slice, index, char::is_whitespace).unwrap_or(index);
+        let category = index.category(slice).unwrap_or(Category::Unknown);
+        backwards_skip_while(slice, index, |c| categorize(c) == category)
+            .map(|i| i + 1)
+            .unwrap_or(0)
+    })
```
Comment (pickfire) [inline helix-core/src/words.rs:16]: We have this pattern quite often. I wonder if we can shorten it to something like `xxx.repeat(count)`?

File: helix-core/src/movement.rs (hunk 6 above)
Comment (pickfire) [reply helix-core/src/movement.rs:155]: I think flipping it will make it easier to understand. I mean rather `impl for usize`, `impl for Rope` will result in slightly longer code but the flow will be more understandable. Rather than `wd`, we have `dw`.

File: helix-core/src/movement.rs (hunk 4 above)
Comment (pickfire) [reply helix-core/src/movement.rs:89]: > As for the ?, it's needed to break out of the closure with a None when we've reached the end of the slice, and thus don't want to make any further anchor moves. This ? is specifically what fixes the previous bug of moving the anchor when already having selected the last word.

Why not rather than using `?`, we `unwrap_or(range)` and just return `Range`?

File: helix-core/src/movement.rs (hunk 5 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:165]: Of course!: https://godbolt.org/z/vMv5x79oP

The compiler eats this kind of redundancy for breakfast. Not just functions or modules, but even across crate boundaries, especially if you enable LTO. That's why I'm never an advocate for marking functions `inline` and other optimizer hints unless you're deep into optimization and looking at the results with a disassembler/profiler.

File: helix-core/src/movement.rs (hunk 4 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:89]: > > As for the ?, it's needed to break out of the closure with a None when we've reached the end of the slice, and thus don't want to make any further anchor moves. This ? is specifically what fixes the previous bug of moving the anchor when already having selected the last word.
> 
> Why not rather than using `?`, we `unwrap_or(range)` and just return `Range`?

I like that, it might be more readable that way. Gonna check that it doesn't break anything but it looks good :+1:

File: helix-core/src/movement.rs [hunk 8]
```diff
@@ -1,184 +1,254 @@
...
 
-        // refetch
-        let ch = slice.char(end);
+/// Private helpers to help manipulate slice indices
+pub trait SliceIndexHelpers {
+    fn outside(&self, slice: RopeSlice) -> bool;
+    fn inside(&self, slice: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_boundary(&self, slice: RopeSlice) -> bool;
```
Comment (pickfire) [inline helix-core/src/movement.rs:141]:
```suggestion
    fn next_char_is_boundary(&self, slice: RopeSlice) -> bool;
```
I think this can make it more obvious.

File: helix-core/src/movement.rs [hunk 9]
```diff
@@ -1,184 +1,254 @@
...
+    fn category(&self, slice: RopeSlice) -> Option<Category>;
+    /// Returns the start of a word/punctuation group followed by any amount of whitespace.
```
Comment (pickfire) [inline helix-core/src/movement.rs:143]: I don't quite understand this. Followed by means on the right right?

File: helix-core/src/movement.rs [hunk 10]
```diff
@@ -1,184 +1,254 @@
...
+    fn category(&self, slice: RopeSlice) -> Option<Category> {
+        self.inside(slice).then(|| categorize(slice.char(*self)))
+    }
 
-        if categorize(ch) != categorize(next) {
-            begin += 1;
-        }
+    fn end_of_word(&self, slice: RopeSlice) -> Self {
+        // Scan the entire slice
+        (*self..slice.len_chars())
```
Comment (pickfire) [inline helix-core/src/movement.rs:175]: Why not use `chars_at(self)` with their iterator API?

File: helix-core/src/movement.rs [hunk 11]
```diff
@@ -1,184 +1,254 @@
...
+    fn end_of_block(&self, slice: RopeSlice) -> Self {
+        // Scan the entire slice
+        (*self..slice.len_chars())
+            // Skip any initial newlines, as they must be skipped over for
+            // the purposes of word movement
+            .skip_while(|pos| pos.is_end_of_line(slice))
+            // Find the first boundary that doesn't go into whitespace or EOL
+            .find(|pos| {
+                pos.is_boundary(slice)
+                    &&((pos + 1).is_end_of_line(slice) || !slice.char(*pos + 1).is_whitespace())
```
Comment (pickfire) [inline helix-core/src/movement.rs:197]: We are getting `char` too many times when we only need to get it twice.

File: helix-core/src/movement.rs [hunk 12]
```diff
@@ -1,184 +1,254 @@
...
+    fn end_of_word(&self, slice: RopeSlice) -> Self {
+        // Scan the entire slice
+        (*self..slice.len_chars())
+            // Skip any initial newlines, as they must be skipped over for
+            // the purposes of word movement
+            .skip_while(|i| is_end_of_line(slice.char(*i)))
+            // Find the first boundary that doesn't start from whitespace
+            .find(|pos| {
+                pos.is_boundary(slice)
+                    && (!pos.is_whitespace(slice) || (pos + 1).is_end_of_line(slice))
```
Comment (pickfire) [inline helix-core/src/movement.rs:182]: I think this only needs two `.char` but we did it too many times here. `.char` is `O(log N)` and if rust compiler cannot optimize this means we are getting 2x unnecessary calls.

File: helix-core/src/movement.rs [hunk 13]
```diff
@@ -273,17 +369,445 @@ mod test {
...
+
+    enum Motion {
+        NextStart(usize),
+        NextEnd(usize),
+        PrevStart(usize),
+    }
+
+    #[test]
+    fn test_behaviour_when_moving_to_start_of_next_words() {
+        let tests = array::IntoIter::new([
```
Comment (pickfire) [inline helix-core/src/movement.rs:547]: Can we just use `vec!` here? We don't really need iterator here, it's tests anyway, having it simpler would be better.

**Comment by pickfire** [comment]
Good job on the tests. I really like that, very comprehensive.

Look at most parts but didn't really go through every single bit.

But on the new functional style, I find it a bit weird to see chaining of usize, I think it makes it a bit unreadable and may be confusing for new comers, the functional parts on iterator is still fine, just the `impl X for usize` makes it weird. I think it would be better if that is reversed to break the chaining so you can do `let x = slice.function(x);` and split it across lines. The only good thing is it is shorter but not really easier to understand (at least for me).

Another thing I am concerned is whether rust have the ability optimized it, I see that some function calls are repeated quite a few times even though it is only necessary to be checked once, one way to solve those is just to use `debug_assert!` but I don't think that works here. Another thing is I think we are straying from `chars_at` and find ourselves using more `char`, `chars_at` should be preferred in the long run because it runs at an amortized `O(1)`, good for frequently run stuff, but if we do `char`, and if rust compiler could not optimize it we are leaving performance at the table.

File: helix-core/src/movement.rs (hunk 5 above)
Comment (pickfire) [reply helix-core/src/movement.rs:173]: Yeah, the simpler ones should be possible. I am just a bit afraid of the larger ones.

File: helix-term/src/commands.rs [hunk 14]
```diff
@@ -2198,7 +2153,7 @@ pub fn join_selections(cx: &mut Context) {
         for line in lines {
             let mut start = text.line_to_char(line + 1).saturating_sub(1);
             let mut end = start + 1;
-            skip_over_next(slice, &mut end, |ch| matches!(ch, ' ' | '\t'));
+            end = skip_while(slice, end, |ch| matches!(ch, ' ' | '\t')).unwrap_or(end);
```
Comment (kirawi) [inline helix-term/src/commands.rs:2156]: Ah, I think I may have missed this in #135. This should be `is_strict_whitespace` instead of `|ch| matches!(ch, ' ' | '\t'`

File: helix-core/src/movement.rs (hunk 5 above)
Comment (pickfire) [reply helix-core/src/movement.rs:173]: https://godbolt.org/z/rYYPrzxs5 I think it still does compare

Comment (PabloMansanet) [reply helix-core/src/movement.rs:173]: It's an interesting point, and one that pertains to the approach to how the entire project is written.

In general, most developers' intuition of what the compiler can do falls *very, very short*. If you hold immutable references to data through an entire process, no matter how many functions, modules or even crates it spans, you have a near certainty that any redundancy will be optimized away. In fact, what's likely is that redundancies and inefficiencies you hadn't even considered will get optimized away (think things reordering, branch elimination, loop unrolling, etc).

My personal approach is usually not to worry about optmization until it's time to optimize, and when it's time to optimize to do it with proper benchmarking, looking at the generated assembly, identifying performance bottlenecks, etc. The risk of premature optimization over perceived, source-level inefficiencies is that it has a risk of unnecessarily complicating both the work and the future refactors, chasing performance problems that may not even be there.

It would be good for @archseer to establish a baseline for the project when it comes to performance and optimization, so we know what the expectation is.

File: helix-core/src/movement.rs (hunk 13 above)
Comment (archseer) [reply helix-core/src/movement.rs:556]: I mentioned this before, in the next stable compiler version we can simply drop `IntoIter` and keep the array.

File: helix-core/src/movement.rs [hunk 15]
```diff
@@ -1,184 +1,257 @@
...
+        self.skip_while(|(_, c)| is_end_of_line(*c))
+    }
 }
 
 // ---- util ------------
-
-// used for by-word movement
+/// Returns a forward and backwards iterator over (usize, char), where the first element
+/// always corresponds to the absolute index of the character in the slice.
+pub fn enumerated_chars<'a>(
```
Comment (pickfire) [inline helix-core/src/movement.rs:224]: I see the usage for this function is always only one of the returned output is used, I think we can split this into two functions, one for forward and one for backward.

File: helix-core/src/movement.rs [hunk 16]
```diff
@@ -1,184 +1,257 @@
...
-
-        if !skip_over_next(slice, &mut begin, |ch| ch == '\n') {
-            return None;
+pub fn move_next_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let movement = |range: Range| -> Result<Range, Range> {
+        let (characters, _) = enumerated_chars(&slice, range.head);
+        let new_head = characters.clone().skip(1).end_of_block().ok_or(range)?;
+        let new_anchor = if characters.clone().at_boundary() {
+            characters
+                .clone()
```
Comment (pickfire) [inline helix-core/src/movement.rs:92]: I see a lot of `clone()` here, since we are using an iterator based API, maybe we can use `by_ref()` instead?

File: helix-core/src/movement.rs (hunk 5 above)
Comment (pickfire) [reply helix-core/src/movement.rs]: Yes, usually I don't worry about optimization but I worry about adding needless constructs that does a lot more than it need to. In this case which may potentially slows down performance. If we already always add those beforehand, we can just add a `debug_assert!` here just to double confirm, I don't think we need the full check.

**Review by pickfire — COMMENTED** [review]
New version now looks more readable.

I haven't been able to look at it closely but I see that it can be improved by replacing those `clone()` with `by_ref()` and `peek()`.

I also commented on a function that only one part of the tuple returned will be used which I think we can split it out.

Some parts like the `at_boundary() { ... } new_anchor` part seemed quite duplicated, I wonder if we can slim those parts.

Maybe tonight I will take a closer look and test it out as well.
